working/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,382 @@
+
+Status: In-progress
+
+Version 0.1 (see [CHANGELOG](#CHANGELOG) at end)
+
+Experiment flag: private-named-parameters
+
+This proposal makes it easier to initialize and declare private instance fields
+using named constructors parameters. It addresses [#2509][] and turns code like
+this:
+
+[#2509]: https://github.com/dart-lang/language/issues/2509
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({
+    int? windows,
+    int? bedrooms,
+    int? swimmingPools,
+  })  : _windows = windows,
+        _bedrooms = bedrooms,
+        _swimmingPools = swimmingPools;
+}
+```
+
+Into this:
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({this._windows, this._bedrooms, this._swimmingPools});
+}
+```
+
+## Motivation
+
+Dart uses a leading underscore in an identifier to make a declaration private to
+its library. Privacy is only meaningful for declarations that could be accessed
+from outside of the library: top-level declarations and members on types.
+
+Local variables and parameters aren't in scope outside of the library where they
+are defined, so privacy doesn't come into play. Except, that is, for named
+parameters. A *named* parameter has one foot on each side of the function
+boundary. The parameter defines a local variable that is accessible inside the
+function, but it also specifies the name used at the callsite to pass an
+argument for that parameter:
+
+```dart
+test({String? _hmm}) {
+  print(_hmm);
+}
+
+main() {
+  test(_hmm: 'ok?');
+}
+```
+
+A public function containing a named parameter whose name is private raises
+difficult questions. Is there any way to pass an argument to the function from
+outside of the library? If the parameter is required, does that mean the
+function is effectively uncallable? Or do we not treat the identifier as private
+even though it starts with an underscore if it happens to be a parameter name?
+
+The language currently resolves these questions by routing around them: it is a
+compile-time error to have a named parameter with a private name. Users must use
+a public name instead. For most named parameters, this restriction is harmless.
+The parameter is only used within the body of the function and its idiomatic for
+local variables to not have private names anyway.
+
+### Initializing formals
+
+However, initializing formals (the `this.` before a constructor parameter)
+complicate that story. When a named parameter is also an initializing formal,
+then the name affects *three* places in the program:
+
+1.  The name of the parameter variable inside the body of the constructor.
+
+2.  The name used to pass an argument at the callsite.
+
+3.  The name of the corresponding instance field to initialize with that
+    parameter.
+
+For example:
+
+```dart
+class House {
+  int? bedrooms; // 3. The corresponding field.
+  House({this.bedrooms}) {
+    print(bedrooms); // 1. The parameter variable.
```

Comment:
Actually references the instance variable. The `this.bedrooms` parameter only introduce a variable in the initializer list.

So, an example would be:
```dart
class House {
  int? bedrooms; // 3. The corresponding field.
  House({this.bedrooms}) 
      : assert(bedrooms == null || bedrooms >= 0); // 1. The parameter variable
}
```

Reply:
Oh, good catch. Done.